pkg/envoy/endpoint.go

```diff
@@ -0,0 +1,167 @@
+type versionBinding struct {
+	cacheVersion   uint64
+	policyRevision uint64
+	endpoint       NetworkPolicyEndpoint
+}
+
+// Insert an {endpoint, policy revision, NPDS cache version} binding into the
+// list so that NetworkPolicyUpdateObservers may find and update the proxy
+// policy revision associated with an endpoint.
+func (npel *NetworkPolicyEventList) Insert(ep NetworkPolicyEndpoint, policyRevision, cacheVersion uint64) {
+	addrs := []string{
+		ep.GetIPv4Address(),
+		ep.GetIPv6Address(),
+	}
+
+	npel.Lock()
+	defer npel.Unlock()
+	for _, addr := range addrs {
+		if addr == "" {
+			continue
+		}
+
+		value := versionBinding{
+			cacheVersion:   cacheVersion,
+			policyRevision: policyRevision,
+			endpoint:       ep,
+		}
+		revisions := npel.ipToBinding[addr]
+		npel.ipToBinding[addr] = append(revisions, value)
+	}
+}
+
+// Remove an endpoint from the list so that it no longer receives callbacks
+// when the NetworkPolicy is successfully updated.
+func (npel *NetworkPolicyEventList) Remove(ep NetworkPolicyEndpoint) {
+	addrs := []string{
+		ep.GetIPv4Address(),
+		ep.GetIPv6Address(),
+	}
+
+	npel.Lock()
+	defer npel.Unlock()
+	for _, addr := range addrs {
+		if addr != "" {
+			delete(npel.ipToBinding, addr)
+		}
+	}
+}
+
+// getRevisionBinding fetches the policy revision and endpoint pointer
+// corresponding to the specified address and NPDS cache version.
+func (npel *NetworkPolicyEventList) getRevisionBinding(addr string, version uint64) (policyRevision uint64, ep NetworkPolicyEndpoint) {
+	npel.Lock()
+	defer npel.Unlock()
+
+	revisions := npel.ipToBinding[addr]
+	newRevisions := make([]versionBinding, 0, len(revisions))
+	for _, r := range revisions {
+		if r.cacheVersion > version {
+			newRevisions = append(newRevisions, r)
+		} else if r.cacheVersion == version {
+			policyRevision = r.policyRevision
+			ep = r.endpoint
+		}
+	}
+	// Clean up all revisions that are no longer interesting (ie, older
+	// than the version currently being acknowledged)
+	npel.ipToBinding[addr] = newRevisions
+
+	return
+}
+
+// NewNetworkPolicyEventList creates an empty new event list.
+func NewNetworkPolicyEventList() *NetworkPolicyEventList {
+	return &NetworkPolicyEventList{
+		ipToBinding: make(map[string][]versionBinding, defaultipToBindingSize),
+	}
+}
+
+// networkPolicyUpdateObserver handles resource acknowledgements for the
+// NetworkPolicyDiscoveryService.
+type networkPolicyUpdateObserver struct{}
+
+// HandleResourceVersionAck pushes the acknowledged policy revision into the
+// endpoint.
+func (o *networkPolicyUpdateObserver) HandleResourceVersionAck(version uint64, node *envoy_api_v2_core.Node, resourceNames []string, typeURL string) {
+	if typeURL != NetworkPolicyTypeURL {
+		return
+	}
+
+	addr, err := xds.IstioNodeToIP(node)
+	if err != nil {
+		log.Debugf("Recieved invalid address from Istio: %v", node)
+		return
+	}
+
+	policyRevision, ep := unackedNPDSUpdates.getRevisionBinding(addr, version)
```

Comment:
We need to have a different behavior depending on whether we run Envoy on the host or in sidecar proxies, based on the `sidecar-http-proxy` flag (https://github.com/cilium/cilium/pull/3126).
In case that flag is false, the node IP address will always be `127.0.0.1`, because all endpoints are handled by the local Envoy proxy. In that case, you need to "wildcard" the node address in `getRevisionBindings`, i.e. only match bindings based on cache version.